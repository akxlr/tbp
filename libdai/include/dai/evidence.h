/*  This file is part of libDAI - http://www.libdai.org/
 *
 *  Copyright (c) 2006-2011, The libDAI authors. All rights reserved.
 *
 *  Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
 */


/// \file
/// \brief Defines class Evidence, which stores multiple observations of joint states of variables


#ifndef __defined_libdai_evidence_h
#define __defined_libdai_evidence_h


#include <istream>
#include <dai/daialg.h>


namespace dai {


/// Stores a data set consisting of multiple samples, where each sample is the observed joint state of some variables.
/** \note Each sample can describe the joint state of a different set of variables,
 *  in order to be able to deal with missing data.
 *
 *  \author Charles Vaske
 */
class Evidence {
    public:
        /// Stores joint state of a set of variables
        typedef std::map<Var, size_t> Observation;

    private:
        /// Each sample is an observed joint state of some variables
        std::vector<Observation> _samples;

    public:
        /// Default constructor
        Evidence() : _samples() {}

        /// Construct from \a samples
        Evidence( std::vector<Observation> &samples ) : _samples(samples) {}

        /// Read in tabular data from a stream and add the read samples to \c *this.
        /** \param is Input stream in .tab file format, describing joint observations of variables in \a fg
         *  \param fg Factor graph describing the corresponding variables
         *  \see \ref fileformats-evidence
         *  \throw INVALID_EVIDENCE_FILE if the input stream is not valid
         */
        void addEvidenceTabFile( std::istream& is, FactorGraph& fg );

        /// Returns number of stored samples
        size_t nrSamples() const { return _samples.size(); }

    /// \name Iterator interface
    //@{
        /// Iterator over the samples
        typedef std::vector<Observation>::iterator iterator;
        /// Constant iterator over the samples
        typedef std::vector<Observation>::const_iterator const_iterator;

        /// Returns iterator that points to the first sample
        iterator begin() { return _samples.begin(); }
        /// Returns constant iterator that points to the first sample
        const_iterator begin() const { return _samples.begin(); }
        /// Returns iterator that points beyond the last sample
        iterator end() { return _samples.end(); }
        /// Returns constant iterator that points beyond the last sample
        const_iterator end() const { return _samples.end(); }
    //@}

    private:
        /// Read in tabular data from a stream and add the read samples to \c *this.
        void addEvidenceTabFile( std::istream& is, std::map<std::string, Var> &varMap );
};


} // end of namespace dai


#endif
