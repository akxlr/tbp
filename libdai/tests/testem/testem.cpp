/*  This file is part of libDAI - http://www.libdai.org/
 *
 *  Copyright (c) 2006-2011, The libDAI authors. All rights reserved.
 *
 *  Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
 */


#include <iostream>
#include <fstream>
#include <string>

#include <dai/factorgraph.h>
#include <dai/evidence.h>
#include <dai/alldai.h>


using namespace std;
using namespace dai;


void usage( const string &msg ) {
    cerr << msg << endl;
    cerr << "Usage:" << endl;
    cerr << " testem factorgraph.fg evidence.tab emconfig.em" << endl;
    exit( 1 );
}


int main( int argc, char** argv ) {
    if( argc != 4 )
        usage("Incorrect number of arguments.");

    FactorGraph fg;
    fg.ReadFromFile( argv[1] );

    PropertySet infprops;
    infprops.set( "verbose", (size_t)0 );
    infprops.set( "updates", string("HUGIN") );
    InfAlg* inf = newInfAlg( "JTREE", fg, infprops );
    inf->init();

    Evidence e;
    ifstream estream( argv[2] );
    e.addEvidenceTabFile( estream, fg );

    cout << "Number of samples: " << e.nrSamples() << endl;
    for( Evidence::iterator ps = e.begin(); ps != e.end(); ps++ )
        cout << "Sample #" << (ps - e.begin()) << " has " << ps->size() << " observations." << endl;

    ifstream emstream( argv[3] );
    EMAlg em(e, *inf, emstream);

    while( !em.hasSatisfiedTermConditions() ) {
        Real l = em.iterate();
        cout << "Iteration " << em.Iterations() << " likelihood: " << l <<endl;
    }

    cout << endl << "Inferred Factor Graph:" << endl << "######################" << endl;
    cout.precision(12);
    cout << inf->fg();

    delete inf;

    return 0;
}
