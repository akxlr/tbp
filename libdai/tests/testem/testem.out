Number of samples: 20
Sample #0 has 2 observations.
Sample #1 has 2 observations.
Sample #2 has 2 observations.
Sample #3 has 2 observations.
Sample #4 has 2 observations.
Sample #5 has 2 observations.
Sample #6 has 2 observations.
Sample #7 has 2 observations.
Sample #8 has 2 observations.
Sample #9 has 2 observations.
Sample #10 has 2 observations.
Sample #11 has 2 observations.
Sample #12 has 2 observations.
Sample #13 has 2 observations.
Sample #14 has 2 observations.
Sample #15 has 2 observations.
Sample #16 has 2 observations.
Sample #17 has 2 observations.
Sample #18 has 2 observations.
Sample #19 has 2 observations.
Iteration 1 likelihood: -27.7259
Iteration 2 likelihood: -23.4297
Iteration 3 likelihood: -23.4297

Inferred Factor Graph:
######################
1

2
0 1 
2 2 
4
0   0.166666666667
1   0.666666666667
2   0.833333333333
3   0.333333333333
Number of samples: 20
Sample #0 has 2 observations.
Sample #1 has 2 observations.
Sample #2 has 2 observations.
Sample #3 has 2 observations.
Sample #4 has 2 observations.
Sample #5 has 2 observations.
Sample #6 has 2 observations.
Sample #7 has 2 observations.
Sample #8 has 2 observations.
Sample #9 has 2 observations.
Sample #10 has 2 observations.
Sample #11 has 2 observations.
Sample #12 has 2 observations.
Sample #13 has 2 observations.
Sample #14 has 2 observations.
Sample #15 has 2 observations.
Sample #16 has 2 observations.
Sample #17 has 2 observations.
Sample #18 has 2 observations.
Sample #19 has 2 observations.
Iteration 1 likelihood: -27.7259
Iteration 2 likelihood: -23.7555
Iteration 3 likelihood: -23.7555

Inferred Factor Graph:
######################
1

3
0 1 2 
2 2 2 
8
0   0.214285714286
1   0.642857142857
2   0.785714285714
3   0.357142857143
4   0.214285714286
5   0.642857142857
6   0.785714285714
7   0.357142857143
Number of samples: 5
Sample #0 has 5 observations.
Sample #1 has 4 observations.
Sample #2 has 6 observations.
Sample #3 has 6 observations.
Sample #4 has 5 observations.
Iteration 1 likelihood: -16.4893
Iteration 2 likelihood: -15.4406
Iteration 3 likelihood: -15.3299

Inferred Factor Graph:
######################
3

3
2 6 7 
2 2 2 
8
0   0.396387112396
1   0.356649908835
2   0.603612887604
3   0.643350091165
4   0.805779039527
5   0.669278114097
6   0.194220960473
7   0.330721885903

3
0 1 6 
2 2 2 
8
0    1.03521336269
1    1.54747895212
2    2.31765218974
3    1.28041900719
4      4.922079813
5    2.52725575019
6   0.831279296316
7   0.262805630803

3
1 2 4 
2 2 2 
8
0   0.396387112396
1   0.603612887604
2   0.356649908835
3   0.643350091165
4   0.805779039527
5   0.194220960473
6   0.669278114097
7   0.330721885903
Number of samples: 5
Sample #0 has 5 observations.
Sample #1 has 4 observations.
Sample #2 has 6 observations.
Sample #3 has 6 observations.
Sample #4 has 5 observations.
Iteration 1 likelihood: -16.4893
Iteration 2 likelihood: -17.6905
Iteration 3 likelihood: -17.6582

Inferred Factor Graph:
######################
3

3
2 6 7 
2 2 2 
8
0   0.495312199726
1   0.499107942908
2   0.504687800274
3   0.500892057092
4   0.640416549958
5   0.495122293924
6   0.359583450042
7   0.504877706076

3
0 1 6 
2 2 2 
8
0   0.495312199726
1   0.504687800274
2   0.499107942908
3   0.500892057092
4   0.640416549958
5   0.359583450042
6   0.495122293924
7   0.504877706076

3
1 2 4 
2 2 2 
8
0   0.495312199726
1   0.504687800274
2   0.499107942908
3   0.500892057092
4   0.640416549958
5   0.359583450042
6   0.495122293924
7   0.504877706076
